File: Sliding window/1838. Frequency of the Most Frequent Element.cpp
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k) {
        int res = 1, start = 0;
        long cur = 0;
        sort(nums.begin(), nums.end());
        for(int i = 1; i<nums.size(); ++i){
            cur += static_cast<long>( (nums[i] - nums[i-1])) * (i-start); //把之前每个数都增加到 nums[i]
            while(cur > k && start <= i){
                cur -= (nums[i] - nums[start++]);
            }
            res = max(res, i - start + 1);
        }
        return res;
    }
};

//2025
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int l = 0; 
        long diff = 0;
        int res = 0;
        for(int i = 0; i < n; ++i){
            if(i > 0 && nums[i] != nums[i-1])
                diff += (long)(nums[i]-nums[i-1])*(i-l);
            while (diff > k) {
                diff -= nums[i] - nums[l++];
            }
            res = max(res, i-l+1);
        }
        return res;
    }
};

//2025
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int l = 0; 
        long diff = 0;
        int res = 0;
        for(int i = 0; i < n; ++i){
            if(i > 0 && nums[i] != nums[i-1])
                diff += (long)(nums[i]-nums[i-1])*(i-l);
            if (diff > k) {
                diff -= nums[i] - nums[l++];
            }
        }
        return n-l;
    }
};
/*

valid windows

k  >= windows_size * max - sum

k + sum >= windows_size * max


For every new element A[j] to the sliding window,
Add it to the sum by sum += A[j].
Check if it'a valid window by
sum + k < (long)A[j] * (j - i + 1)

*/


// (Non-shrinkable)
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k) {
        int i = 0, j;
        sort(nums.begin(), nums.end());
        for (j = 0; j < nums.size(); ++j) {
            k += nums[j];
            if (k < (long)nums[j] * (j - i + 1))
                k -= nums[i++];
        }
        return j - i;
    }
};

//(Shrinkable)
class Solution {
public:
    int maxFrequency(vector<int>& A, int k) {
        sort(begin(A), end(A));
        long i = 0, N = A.size(), ans = 1, sum = 0;
        for (int j = 0; j < N; ++j) {
            sum += A[j];
            while ((j - i + 1) * A[j] - sum > k) sum -= A[i++];
            ans = max(ans, j - i + 1);
        }
        return ans;
    }
};

File: Sliding window/1838. Frequency of the Most Frequent Element.py
class Solution:
    def maxFrequency(self, nums: List[int], k: int) -> int:
        nums.sort()
        res = 1
        cur = 0
        start = 0
        for i, v in enumerate(nums[1:],1):
            cur += (i - start)*(v-nums[i-1])
            while cur > k and start<=i and start < len(nums):
                cur -= v - nums[start]
                start += 1
            res = max(res, i- start+1)
        return res

"""
valid windows

k  >= windows_size * max - sum

k + sum >= windows_size * max
"""

class Solution:
    def maxFrequency(self, nums: List[int], k: int) -> int:
        i = 0
        nums.sort()
        for j in range(len(nums)):
            k += nums[j]
            if k < nums[j] * (j - i + 1):
                k -= nums[i]
                i += 1
        return j - i + 1


class Solution:
    def maxFrequency(self, nums: List[int], k: int) -> int:
        n = len(nums)
        nums.sort()
        preSum = [0] * (n + 1)
        for i in range(n):
            preSum[i + 1] = preSum[i] + nums[i]

        def getSum(left, right):  # left, right inclusive
            return preSum[right + 1] - preSum[left]

        def count(index): # Count frequency of `nums[index]` if we make other elements equal to `nums[index]`
            left = 0
            right = index
            res = index
            while left <= right:
                mid = left + (right - left) // 2
                s = getSum(mid, index - 1) # get sum of (nums[mid], nums[mid+1]...nums[index-1])
                if s + k >= (index - mid) * nums[index]: # Found an answer -> Try to find a better answer in the left side
                    res = mid  # save best answer so far
                    right = mid - 1
                else:
                    left = mid + 1
            return index - res + 1

        ans = 0
        for i in range(n):
            ans = max(ans, count(i))
        return ans